lilcommodule.c

```diff
@@ -1,39 +1,307 @@
+  int n_dims = 0; // Number of dimensions of the given numpy array
+  int input_stride; // The number of integers between to consecutive samples
+  int output_stride; // The number of integers between to consecutive samples in output
+  int n_samples = 0; // number of samples given in the numpy array
+  // INPUT IS DEFINED INSIDE IF CONDITIONS
+  int8_t *output; // The one dimensional vectorized output array which will be modified by the core function
+  int lpc_order = 5; // LPC Order defined in the core function (more information -> lilcom.h)
+  int integral = 1; // Checks whether the array contains integer or float
+
+  /* Reading and information - extracting for input data */
+  static char *kwlist[] = {"X" , "lpc_order", NULL};
+  if (!PyArg_ParseTupleAndKeywords(args, keywds, "O|i", kwlist, &signal_object, &lpc_order)) 
+    Py_RETURN_FALSE;
+  
+  // Initializing shape related variables
+  n_dims = PyArray_NDIM(signal_object); // Getting the number of dimensions
+  n_samples = PyArray_DIM(signal_object , 0); // Getting the first dimension
+  if (n_dims > 1) input_stride = PyArray_DIM(signal_object , 0); // Conditioning on second dimension
+  else input_stride = 1;
+  integral = integral_check(((PyArrayObject*)signal_object)->descr->type_num); // Checking whether the array has float
+
+
+  /* Debug: Comment or Uncomment when on debug */
+  // printf("The array is received and the lpc_order = %d\n", lpc_order);
+  // printf("Array is %d dimensional\n", n_dims);
+  // printf("It has %d samples and stride is %d\n", n_samples, input_stride);
+  // printf("PyArrayType == int is %d\n", integral );
+
+
+  output_stride = input_stride; ////// CHANGE IT IF NEEDED
+
+  void *signal_object_data = PyArray_DATA(signal_object);
+
+  if(integral == 1) { // Integer
+    int16_t *input; // The one dimensional vectorized input array which will be given to the core function
+    /* Allocating the space for input and output */
+    input = malloc(sizeof(int16_t) * n_samples * input_stride);
+    output = malloc(sizeof(int8_t) * n_samples * output_stride);
+    
+
+    /* Conversion to int16_t array */
+    switch (((PyArrayObject*)signal_object)->descr->type_num){
+      case NPY_INT8:
+        for (int i = 0 ; i < n_samples * input_stride ; i++){
+          input[i] = (int16_t)(((int8_t *)signal_object_data)[i]); 
+        } break;
+      case NPY_INT16:
+        for (int i = 0 ; i < n_samples * input_stride ; i++){
+          input[i] = (int16_t)(((int16_t *)signal_object_data)[i]); 
+        } break;
+      case NPY_INT32:
+        for (int i = 0 ; i < n_samples * input_stride ; i++){
+          input[i] = (int16_t)(((int32_t *)signal_object_data)[i]); 
+        } break;
+      case NPY_INT64:
+        for (int i = 0 ; i < n_samples * input_stride ; i++){
+          input[i] = (int16_t)(((int64_t *)signal_object_data)[i]); 
+        } break;
+      case NPY_UINT8:
+        for (int i = 0 ; i < n_samples * input_stride ; i++){
```

Comment:
this code seems to assume that the array is contiguous.  We shouldn't assume that, as far as the input array is concerned.  (Of course, we have control of any temporary array).
In any case, I don't think it's necessary to support types other than int16 at the current point.  Just casting wouldn't be right anyway, e.g. from int32 you'd need to correct for the different range.   I want to support float input too, but it requires some changes to the core code to make the scale representable.

Reply:
I thought that the array we receive is contiguous as we receive it in form of a numpy array, I'll cover that. Also the reason of castings and the switch case is that the input may be given as a numpy array of any form. At last what we have is a 16-bit integer (with its proper range), but when you make a numpy array, there are cases (like reading from some kind of csv, making copies from list or ...) that the array is automatically formed in different integer format. This is just to ensure that the data given to the core function is correctly formatted.